### ISSUE TITLE
Class structure for fitting

### ISSUE THREAD

**Issue by andrey-popov** [IPNL-CMS/multijet-jec#12]
JEC fit is to be reimplemented for [method&nbsp;B](https://indico.cern.ch/event/656050/#65-comparison-of-different-app). Develop a class structure that would be suitable to incorporate eventually measurements in other topologies than multijet. The computation is moderately CPU-intensive, especially in multijets, and therefore should be implemented in C++.

**Comment by andrey-popov** [comment]
The interface to compute the deviation between data and simulation is implemented with class `DeviationBase`, which provides a pure virtual method
```C++
double DeviationBase::Eval(vector<double> const &params, vector<double> const &nuisances)
```
Here `params` are parameters of the jet correction, and `nuisances` are values for nuisances. The actual computation is to be implemented in a derived class, which would be specific to each topology. Should probably also define a method to report the number of degrees of freedom. At the construction time this class also saves a reference to an object that evaluates the jet correction.

The fit is to be performed by minimizing the sum of deviations computed in each topology. Might need to allow relative weights. The sum is to be implemented with a dedicated class, which would also provide an interface to interact with the minimizer (like method `operator()(double const *)` for Minuit).

The jet correction is to be implemented in a class inheriting from `JetCorrBase`, which defines a pure virtual method
```C++
double JetCorrBase::Eval(double pt, double eta, vector<double> const &params)
```
Properties of the jet can be specified in the signature explicitly since they are used in the algorithms to compute the deviations and therefore their list will not change easily. The inclusion of &eta; might be needed for the multijet topology as there likely to be a residual dependence on pseudorapidity after the L2Res corrections. This might also allow a simultaneous determination of L2Res and L3Res corrections. If only the L3Res correction is fitted and thus it does not depend on &eta;, the method can be called with an arbitrary value for &eta;. In future the quark-gluon and b-tagging discriminators can be added into dependencies of the residual corrections, but this won't happen any time soon since such dependencies must first be incorporated into the MC-truth corrections. The correction defined by `JetCorrBase` is assumed to be purely multiplicative; a correction to the direction of the jet's momentum is not in plans and would require a rework of L3Res analyses.

**Comment by andrey-popov** [comment]
Nuisance parameters can be numerous and will be shared by multiple analyses. Because of this, they should be accessed by name rather than an index. Implement in a structure or a wrapper around a vector that also allows access by name.

**Comment by andrey-popov** [comment]
Fixed by commit b111b77562d2df0715a0db5b3c4eb662fcb5a0d3.
